File: cli/src/commands/worktree/done.ts
import { Command } from "commander";
import { daemonPost } from "../../lib/daemon-client.js";
import { preflight } from "../../lib/preflight.js";
import { die, success } from "../../lib/output.js";

export function registerWorktreeDone(worktree: Command): void {
  worktree
    .command("done <id>")
    .description("Mark all agent sessions done and release the worktree's tmux/agent processes")
    .action(async (id: string) => {
      await preflight();
      const result = await daemonPost<{ ok: true; updated: number; terminalsReleased: number }>(
        `/worktrees/${encodeURIComponent(id)}/done`,
      );
      if (!result.ok) die(result.error, result.status === 404 ? 2 : 1);
      const { updated, terminalsReleased } = result.data;
      success(
        `Worktree marked as done: ${id} ` +
          `(${updated} agent session(s) done, ${terminalsReleased} terminal(s) released)`,
      );
    });
}

File: rust/vst-cli/src/commands/worktree/done.rs
//! `vst worktree done <id>`
//!
//! Sends `POST /worktrees/:id/done`. Mirrors `cli/src/commands/worktree/done.ts`.

use vst_types::rest::worktrees::WorktreeDoneResult;

use crate::client::{daemon_post, DaemonResult};
use crate::output::success;
use crate::preflight::preflight;

pub async fn run_worktree_done(id: &str) -> Result<(), (String, i32)> {
    preflight().await;

    let encoded = percent_encode(id);
    let result = daemon_post::<WorktreeDoneResult, ()>(&format!("/worktrees/{encoded}/done"), None)
        .await
        .map_err(|e| (e.to_string(), 1))?;

    match result {
        DaemonResult::Ok { data, .. } => {
            success(&format!(
                "Worktree marked as done: {id} ({} agent session(s) done, {} terminal(s) released)",
                data.updated, data.terminals_released
            ));
            Ok(())
        }
        DaemonResult::Err { status, error, .. } => {
            let code = if status == 404 { 2 } else { 1 };
            Err((error, code))
        }
    }
}

/// Percent-encode characters not safe in URL path segments.
fn percent_encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for byte in s.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(byte as char);
            }
            _ => {
                out.push_str(&format!("%{byte:02X}"));
            }
        }
    }
    out
}
